src/producer.rs

```diff
@@ -1,242 +1,343 @@
+;
+                                },
+                                "json" => {
+                                    if path.file_name().unwrap() == "linked-files-map.json" {
+                                        linkeds.borrow_mut().insert(path.to_str().unwrap().to_string(), self);
+                                    }
+                                },
+                                _ => { }
+                            },
+                            None => { }
+                        }
+                    }
+                }
+            }
+        }
     }
 
-    if let Some(mut gcno_archive) = gcno_archive {
-        for i in 0..gcno_archive.len() {
-            let mut gcno_file = gcno_archive.by_index(i).unwrap();
-            if gcno_file.name() == "linked-files-map.json" {
-                let mut buffer = Vec::new();
-                gcno_file.read_to_end(&mut buffer).unwrap();
-                path_mapping_file = Some(buffer);
-                continue;
+    pub fn read_in_buffer(&self, name: &str, buf: &mut Vec<u8>) -> bool {
+        match *self.item.borrow_mut() {
+            ArchiveType::Zip(ref mut zip) => {
+                let name = name.replace("\\", "/");
+                match zip.borrow_mut().by_name(&name) {
+                    Ok(mut f) => {
+                        f.read_to_end(buf).expect("Failed to read gcda file");
+                        true
+                    },
+                    Err(_) => false,
+                }
+            },
+            ArchiveType::Dir(ref dir) => {
+                match File::open(dir.join(name)) {
+                    Ok(mut f) => {
+                        f.read_to_end(buf).expect("Failed to read gcda file");
+                        true
+                    },
+                    Err(_) => false,
+                }
             }
+        }
+    }
 
-            let gcno_path_in_zip = PathBuf::from(gcno_file.name());
+    pub fn extract(&self, name: &str, path: &PathBuf) -> bool {
+        let dest_parent = path.parent().unwrap();
+        if !dest_parent.exists() {
+            match fs::create_dir_all(dest_parent) {
+                Ok(_) => { },
+                Err(_) => panic!("Cannot create parent directory"),
+            };
+        }
 
-            let path = tmp_dir.join(&gcno_path_in_zip);
+        match *self.item.borrow_mut() {
+            ArchiveType::Zip(ref mut zip) => {
+                let name = name.replace("\\", "/");
+                match zip.borrow_mut().by_name(&name) {
+                    Ok(mut f) => {
+                        let mut file = File::create(&path).expect("Failed to create file");
+                        io::copy(&mut f, &mut file).expect("Failed to copy file from ZIP");
+                        true
+                    },
+                    Err(_) => {
+                        false
+                    },
+                }
+            },
+            ArchiveType::Dir(ref dir) => {
```

Comment:
Why creating symlinks in this case? Can't we parse the files where they are directly?

Reply:
currently, run_gcov expects to have gcno and gcda in the same directory.
So if gcnos are in a zip and gcdas in a directory (or the opposite) then it won't work if we don't create symlinks.

Reply:
Ah right, gcov is a bit of a pain...

Reply:
We can create a hardlink with https://doc.rust-lang.org/std/fs/fn.hard_link.html.

Reply:
why is a hard link better than a soft link ?
Since we never modify gcno/gcda I don't see any interest to have a hard link.

Reply:
It simplifies the code, as it wouldn't require two different implementations for Windows and Linux

Reply:
I just tried to replace soft links by hard ones and tests are failing on my machine because my base dir and the tmp dir are not on the same device.